File: Solutions/131. Palindrome Partitioning/131. Palindrome Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        if(!s.size())
            return res;
        vector<string> path;
        dfs(0,s,path,res);
        return res;
    }
    void dfs(int index,string& s,vector<string>& path,vector<vector<string>>& res){
        if(index==s.size()){
            res.push_back(path);
            return;
        }
        for(int i=index;i<s.size();++i){
            if(foo(s,index,i)){
                path.push_back(s.substr(index,i-index+1));
                dfs(i+1,s,path,res);
                path.pop_back();
            }
        }
    }
    bool foo(const string& s,int start,int end){
        while(start<=end){
            if(s[start++]!=s[end--])
                return false;
        }
        return true;
    }
};

File: Solutions/131. Palindrome Partitioning/131. Palindrome Partitioning.java
class Solution {
    public void foo(List<List<String>> res,List<String> cur,String s,int st){
        if(st==s.length()){
            res.add(new ArrayList(cur));
            return;
        }
        for(int hi=st;hi<s.length();++hi){
            if(isPal(s,st,hi)){
                cur.add(s.substring(st,hi+1));
                foo(res,cur,s,hi+1);
                cur.remove(cur.size()-1);
            }
        }
    }
    public boolean isPal(String s,int st,int ed){
        for(int i=st,j=ed;i<j;++i,--j){
            if(s.charAt(i)!=s.charAt(j)){
                return false;
            }
        }
        return true;
    }
    public List<List<String>> partition(String s) {
        List<List<String>> res=new ArrayList();
        List<String> cur=new ArrayList();
        foo(res,cur,s,0);
        return res;
    }
}